### DISCUSSION TITLE
fastapi v0.109.0: accessing request.client.host in a middleware throws 'NoneType' object has no attribute 'host' when I run a test with pytest

### DISCUSSION THREAD

**Discussion by jbevi** [fastapi/fastapi#11019] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
time_middleware.py:
class TimeMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
       print(f"{request.client.host}")

-------------
test.py:
client = TestClient(app)

def test_ping():
    response = client.get("/v1.0/ping")
    assert response.status_code == 200

------------
main.py:
app= FastAPI(...)
app.add_middleware(paa.TimeMiddleware)
```

### Description

The problem only occurs in v0.109.0 when run tests.
For some reason when running the test, request.client.host is None.
This does not occur when running the fastapi server normally.

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.8.10

**Comment by jbevi** [comment]
The code is an extract of mine.
In the middleware, I should add that the print is after executing: response = await call_next(request)

**Comment by eturino** [comment]
I have the same problem.

**Comment by TechPriest001** [comment]
I do too, I suspect it's related to this change in Starlette mentioned [here](https://stackoverflow.com/a/77787967).

**Comment by eturino** [comment]
it is indeed. I "solved" it by checking if `request.client` existed or not.

**Reply by Kludex** [reply]
This is the right solution.

**Reply by jbevi** [reply]
but the tests will not be able to validate the operations that I perform with the data from request.client.host if I put an if to validate if the data does not come.
It seems more of a workaround than a solution.
I don't just make a print (it was an example) with that information.

**Reply by jbevi** [reply]
Hello, any more considerations or is this patch really considered a solution?

**Reply by jbevi** [reply]
I think it is necessary to reopen this bug to find a real solution.

**Reply by Kludex** [reply]
I already said that is a real solution. You need to check if `client` exists.

**Reply by jbevi** [reply]
The problem is that it does not exist only when running the test (with pytest). So it is obviously a bug that does not exist in previous versions. The proposed "solution" means that functionality must be "evaded" when running the tests, which is not correct.

**Reply by igormoreira-2** [reply]
I am having the same issues when running pytest. It only works with fastapi versions older than 0.105.0. Has anyone here found a solution?

**Comment by sun1638650145** [comment]
When fixing the ReDoS attack on CVE-2024-24762, I encountered the same issue.

**Comment by igormoreira-2** [comment]
I am having the same issues when running pytest. It only works with fastapi versions older than 0.105.0. Has anyone here found a solution?

**Comment by g0lemXIV** [comment]
I was able to resolve the issue by updating the Starlette framework to version 0.37.2

**Reply by ar090** [reply]
This fixed for me as well now we just need fastapi support through starlette  `0.37.2` for this fix -> https://github.com/encode/starlette/pull/2525 
Currently fastapi is maxed at `0.37.0` making my resolver unhappy. 🙏

**Comment by tiangolo** [comment, marked as answer]
This was updated in Starlette, and FastAPI 0.110.1 upgraded Starlette, maybe that fixes the problem, could you try that? :nerd_face:

**Reply by igormoreira-2** [reply]
Yes... version `0.110.1` is working for me now with no issues. Unit tests are passing. Thanks ;)

**Reply by jbevi** [reply]
Brilliant! The incident no longer occurs in v0.110.1, thanks.

**Reply by nadzyah** [reply]
Unfortunately, didn't work for me (or it's just a problem with the linter)
I'm using FastAPI v0.110.1 and Starlette v0.37.2. However, mypy still complains about this error 😢
